File: Online-judge/leetcode/equal-row-and-column-pairs.py
from collections import defaultdict
from typing import List
"""
Md.Shohag <shohag.fks@gmail.com>
https://leetcode.com/problems/equal-row-and-column-pairs/description/
Time complexity O(N^2)
Space complexity O(1)
"""

class Solution:
    def equalPairs(self, grid: List[List[int]]) -> int:
        row_count = defaultdict(int)
        count = 0

        for row in grid:
            row_count[tuple(row)] += 1

        for col in zip(*grid):
            count += row_count[tuple(col)]
        return count

File: Online-judge/leetcode/equal-row-and-column-pairs.cpp
// Time complexity O(n^3)
// Space complexity O(n^1)

class Solution {
public:
    int equalPairs(vector<vector<int>>& grid) {
        int count = 0;
        int n = (int) grid.size();
        for (int row = 0; row < n; row++) {
            vector<int> v1(grid[row]);
            vector<int> v2;

            for (int i = 0; i < n; i++) {
                for (int j = 0; j < n; j++) {
                   v2.push_back(grid[j][i]); 
                }
                if (v1 == v2) count++;
                v2.clear();
            }
        }
        return count;
    }
};